File: src/commands/records.ts
import { Command } from 'commander';
import { createClient } from '../lib/client.js';
import { printOutput, printError, type OutputFormat } from '../lib/output.js';

export function createRecordsCommand(): Command {
  const records = new Command('records').description('Manage records in a table');

  // List records
  records
    .command('list')
    .description('List records from a table')
    .requiredOption('-b, --base <baseId>', 'Base ID')
    .requiredOption('-t, --table <tableIdOrName>', 'Table ID or name')
    .option('-v, --view <view>', 'View name or ID')
    .option('-f, --filter <formula>', 'Filter by formula')
    .option('-s, --sort <field>', 'Sort by field')
    .option('-d, --direction <direction>', 'Sort direction (asc/desc)', 'asc')
    .option('--fields <fields>', 'Comma-separated list of fields to return')
    .option('-n, --page-size <size>', 'Number of records per page', '100')
    .option('-o, --offset <offset>', 'Pagination offset')
    .option('-a, --all', 'Fetch all records (auto-paginate)')
    .option('--format <format>', 'Output format (json/table/csv)', 'json')
    .action(async (options) => {
      try {
        const client = createClient();
        const params = {
          view: options.view,
          filterByFormula: options.filter,
          pageSize: parseInt(options.pageSize),
          offset: options.offset,
          fields: options.fields?.split(','),
          sort: options.sort
            ? [{ field: options.sort, direction: options.direction as 'asc' | 'desc' }]
            : undefined,
        };

        if (options.all) {
          const records = await client.listAllRecords(
            options.base,
            options.table,
            params
          );
          printOutput(records, options.format as OutputFormat);
        } else {
          const response = await client.listRecords(
            options.base,
            options.table,
            params
          );
          printOutput(response, options.format as OutputFormat);
        }
      } catch (error) {
        printError((error as Error).message);
        process.exit(1);
      }
    });

  // Get a single record
  records
    .command('get')
    .description('Get a single record by ID')
    .requiredOption('-b, --base <baseId>', 'Base ID')
    .requiredOption('-t, --table <tableIdOrName>', 'Table ID or name')
    .requiredOption('-r, --record <recordId>', 'Record ID')
    .option('--format <format>', 'Output format (json/table)', 'json')
    .action(async (options) => {
      try {
        const client = createClient();
        const record = await client.getRecord(
          options.base,
          options.table,
          options.record
        );
        printOutput(record, options.format as OutputFormat);
      } catch (error) {
        printError((error as Error).message);
        process.exit(1);
      }
    });

  // Create a record
  records
    .command('create')
    .description('Create a new record')
    .requiredOption('-b, --base <baseId>', 'Base ID')
    .requiredOption('-t, --table <tableIdOrName>', 'Table ID or name')
    .requiredOption('--fields <json>', 'Fields as JSON object')
    .option('--typecast', 'Enable automatic type casting')
    .option('--format <format>', 'Output format (json/table)', 'json')
    .action(async (options) => {
      try {
        const client = createClient();
        const fields = JSON.parse(options.fields);
        const record = await client.createRecord(options.base, options.table, {
          fields,
          typecast: options.typecast,
        });
        printOutput(record, options.format as OutputFormat);
      } catch (error) {
        printError((error as Error).message);
        process.exit(1);
      }
    });

  // Create multiple records
  records
    .command('create-batch')
    .description('Create multiple records (up to 10)')
    .requiredOption('-b, --base <baseId>', 'Base ID')
    .requiredOption('-t, --table <tableIdOrName>', 'Table ID or name')
    .requiredOption('--records <json>', 'Array of records as JSON')
    .option('--typecast', 'Enable automatic type casting')
    .option('--format <format>', 'Output format (json/table)', 'json')
    .action(async (options) => {
      try {
        const client = createClient();
        const recordsData = JSON.parse(options.records);
        const records = recordsData.map((r: { fields: Record<string, unknown> }) => ({
          fields: r.fields || r,
          typecast: options.typecast,
        }));
        const result = await client.createRecords(
          options.base,
          options.table,
          records
        );
        printOutput(result.records, options.format as OutputFormat);
      } catch (error) {
        printError((error as Error).message);
        process.exit(1);
      }
    });

  // Update a record
  records
    .command('update')
    .description('Update a record (partial update)')
    .requiredOption('-b, --base <baseId>', 'Base ID')
    .requiredOption('-t, --table <tableIdOrName>', 'Table ID or name')
    .requiredOption('-r, --record <recordId>', 'Record ID')
    .requiredOption('--fields <json>', 'Fields to update as JSON object')
    .option('--typecast', 'Enable automatic type casting')
    .option('--format <format>', 'Output format (json/table)', 'json')
    .action(async (options) => {
      try {
        const client = createClient();
        const fields = JSON.parse(options.fields);
        const record = await client.updateRecord(
          options.base,
          options.table,
          options.record,
          fields,
          options.typecast
        );
        printOutput(record, options.format as OutputFormat);
      } catch (error) {
        printError((error as Error).message);
        process.exit(1);
      }
    });

  // Update multiple records
  records
    .command('update-batch')
    .description('Update multiple records (up to 10)')
    .requiredOption('-b, --base <baseId>', 'Base ID')
    .requiredOption('-t, --table <tableIdOrName>', 'Table ID or name')
    .requiredOption('--records <json>', 'Array of {id, fields} objects as JSON')
    .option('--typecast', 'Enable automatic type casting')
    .option('--format <format>', 'Output format (json/table)', 'json')
    .action(async (options) => {
      try {
        const client = createClient();
        const recordsData = JSON.parse(options.records);
        const result = await client.updateRecords(
          options.base,
          options.table,
          recordsData.map((r: { id: string; fields: Record<string, unknown> }) => ({
            id: r.id,
            fields: r.fields,
            typecast: options.typecast,
          }))
        );
        printOutput(result.records, options.format as OutputFormat);
      } catch (error) {
        printError((error as Error).message);
        process.exit(1);
      }
    });

  // Replace a record
  records
    .command('replace')
    .description('Replace a record (full replacement)')
    .requiredOption('-b, --base <baseId>', 'Base ID')
    .requiredOption('-t, --table <tableIdOrName>', 'Table ID or name')
    .requiredOption('-r, --record <recordId>', 'Record ID')
    .requiredOption('--fields <json>', 'All fields as JSON object')
    .option('--typecast', 'Enable automatic type casting')
    .option('--format <format>', 'Output format (json/table)', 'json')
    .action(async (options) => {
      try {
        const client = createClient();
        const fields = JSON.parse(options.fields);
        const record = await client.replaceRecord(
          options.base,
          options.table,
          options.record,
          fields,
          options.typecast
        );
        printOutput(record, options.format as OutputFormat);
      } catch (error) {
        printError((error as Error).message);
        process.exit(1);
      }
    });

  // Upsert records
  records
    .command('upsert')
    .description('Upsert records (create or update based on merge fields)')
    .requiredOption('-b, --base <baseId>', 'Base ID')
    .requiredOption('-t, --table <tableIdOrName>', 'Table ID or name')
    .requiredOption('--records <json>', 'Array of records as JSON')
    .requiredOption('--merge-on <fields>', 'Comma-separated field names to merge on')
    .option('--typecast', 'Enable automatic type casting')
    .option('--format <format>', 'Output format (json/table)', 'json')
    .action(async (options) => {
      try {
        const client = createClient();
        const recordsData = JSON.parse(options.records);
        const fieldsToMergeOn = options.mergeOn.split(',');
        const result = await client.upsertRecords(options.base, options.table, {
          records: recordsData.map((r: { fields: Record<string, unknown> }) => ({
            fields: r.fields || r,
          })),
          fieldsToMergeOn,
          performUpsert: { fieldsToMergeOn },
          typecast: options.typecast,
        });
        printOutput(result, options.format as OutputFormat);
      } catch (error) {
        printError((error as Error).message);
        process.exit(1);
      }
    });

  // Delete a record
  records
    .command('delete')
    .description('Delete a record')
    .requiredOption('-b, --base <baseId>', 'Base ID')
    .requiredOption('-t, --table <tableIdOrName>', 'Table ID or name')
    .requiredOption('-r, --record <recordId>', 'Record ID')
    .option('--format <format>', 'Output format (json/table)', 'json')
    .action(async (options) => {
      try {
        const client = createClient();
        const result = await client.deleteRecord(
          options.base,
          options.table,
          options.record
        );
        printOutput(result, options.format as OutputFormat);
      } catch (error) {
        printError((error as Error).message);
        process.exit(1);
      }
    });

  // Delete multiple records
  records
    .command('delete-batch')
    .description('Delete multiple records (up to 10)')
    .requiredOption('-b, --base <baseId>', 'Base ID')
    .requiredOption('-t, --table <tableIdOrName>', 'Table ID or name')
    .requiredOption('--records <ids>', 'Comma-separated list of record IDs')
    .option('--format <format>', 'Output format (json/table)', 'json')
    .action(async (options) => {
      try {
        const client = createClient();
        const recordIds = options.records.split(',');
        const result = await client.deleteRecords(
          options.base,
          options.table,
          recordIds
        );
        printOutput(result.records, options.format as OutputFormat);
      } catch (error) {
        printError((error as Error).message);
        process.exit(1);
      }
    });

  return records;
}

File: cmd/records.go
package cmd

import (
	"encoding/json"
	"fmt"
	"io"
	"os"
	"strings"

	"github.com/andrejkostal/airtable-cli/internal/client"
)

// RecordsCmd queries and manages table records.
type RecordsCmd struct {
	List   RecordsListCmd   `cmd:"" help:"List records from a table."`
	Get    RecordsGetCmd    `cmd:"" help:"Get a single record."`
	Create RecordsCreateCmd `cmd:"" help:"Create record(s)."`
	Update RecordsUpdateCmd `cmd:"" help:"Update a record."`
	Delete RecordsDeleteCmd `cmd:"" help:"Delete record(s)."`
}

// RecordsListCmd lists records from a table.
type RecordsListCmd struct {
	Table        string `arg:"" help:"Table name or ID."`
	Filter       string `short:"f" help:"Filter formula."`
	Sort         string `short:"s" help:"Sort by field."`
	Direction    string `short:"d" default:"asc" enum:"asc,desc" help:"Sort direction."`
	View         string `short:"v" help:"View name or ID."`
	Fields       string `help:"Comma-separated fields to return."`
	All          bool   `short:"a" help:"Fetch all records (auto-paginate)."`
	Limit        int    `short:"n" default:"100" help:"Records per page."`
	WithFieldIds bool   `help:"Return field IDs instead of names."`
}

// Run lists records from a table.
func (c *RecordsListCmd) Run(globals *Globals) error {
	out, cl, baseID := mustBase(globals, "")
	if cl == nil {
		return nil
	}

	var fields []string
	if c.Fields != "" {
		for _, f := range strings.Split(c.Fields, ",") {
			trimmed := strings.TrimSpace(f)
			if trimmed != "" {
				fields = append(fields, trimmed)
			}
		}
	}

	params := client.ListParams{
		Filter:                c.Filter,
		Sort:                  c.Sort,
		Direction:             c.Direction,
		View:                  c.View,
		Fields:                fields,
		PageSize:              c.Limit,
		ReturnFieldsByFieldId: c.WithFieldIds,
	}

	if c.All {
		records, err := cl.ListAllRecords(baseID, c.Table, params)
		if err != nil {
			out.Error(errorCode(err), err.Error())
			return nil
		}
		out.Success(records)
	} else {
		resp, err := cl.ListRecords(baseID, c.Table, params)
		if err != nil {
			out.Error(errorCode(err), err.Error())
			return nil
		}
		out.Success(resp)
	}

	return nil
}

// RecordsGetCmd gets a single record by ID.
type RecordsGetCmd struct {
	Table    string `arg:"" help:"Table name or ID."`
	RecordId string `arg:"" name:"record-id" help:"Record ID."`
}

// Run gets a single record.
func (c *RecordsGetCmd) Run(globals *Globals) error {
	out, cl, baseID := mustBase(globals, "")
	if cl == nil {
		return nil
	}

	record, err := cl.GetRecord(baseID, c.Table, c.RecordId)
	if err != nil {
		out.Error(errorCode(err), err.Error())
		return nil
	}

	out.Success(record)
	return nil
}

// RecordsCreateCmd creates one or more records.
type RecordsCreateCmd struct {
	Table    string `arg:"" help:"Table name or ID."`
	Fields   string `help:"Fields as JSON object or array of objects."`
	Typecast bool   `help:"Enable automatic type casting."`
}

// Run creates record(s) from JSON fields flag or stdin.
func (c *RecordsCreateCmd) Run(globals *Globals) error {
	out, cl, baseID := mustBase(globals, "")
	if cl == nil {
		return nil
	}

	jsonStr, err := readFieldsInput(c.Fields)
	if err != nil {
		out.Error("VALIDATION_ERROR", err.Error())
		return nil
	}

	// Try parsing as array first.
	var arr []map[string]any
	if json.Unmarshal([]byte(jsonStr), &arr) == nil {
		records, err := cl.CreateRecords(baseID, c.Table, arr, c.Typecast)
		if err != nil {
			out.Error(errorCode(err), err.Error())
			return nil
		}
		out.Success(records)
		return nil
	}

	// Otherwise parse as single object.
	var obj map[string]any
	if err := json.Unmarshal([]byte(jsonStr), &obj); err != nil {
		out.Error("VALIDATION_ERROR", fmt.Sprintf("invalid JSON: %v", err))
		return nil
	}

	record, err := cl.CreateRecord(baseID, c.Table, obj, c.Typecast)
	if err != nil {
		out.Error(errorCode(err), err.Error())
		return nil
	}
	out.Success(record)
	return nil
}

// RecordsUpdateCmd updates a single record.
type RecordsUpdateCmd struct {
	Table    string `arg:"" help:"Table name or ID."`
	RecordId string `arg:"" name:"record-id" help:"Record ID."`
	Fields   string `help:"Fields as JSON object."`
	Typecast bool   `help:"Enable automatic type casting."`
}

// Run updates a record.
func (c *RecordsUpdateCmd) Run(globals *Globals) error {
	out, cl, baseID := mustBase(globals, "")
	if cl == nil {
		return nil
	}

	jsonStr, err := readFieldsInput(c.Fields)
	if err != nil {
		out.Error("VALIDATION_ERROR", err.Error())
		return nil
	}

	var fields map[string]any
	if err := json.Unmarshal([]byte(jsonStr), &fields); err != nil {
		out.Error("VALIDATION_ERROR", fmt.Sprintf("invalid JSON: %v", err))
		return nil
	}

	record, err := cl.UpdateRecord(baseID, c.Table, c.RecordId, fields, c.Typecast)
	if err != nil {
		out.Error(errorCode(err), err.Error())
		return nil
	}

	out.Success(record)
	return nil
}

// RecordsDeleteCmd deletes one or more records.
type RecordsDeleteCmd struct {
	Table     string   `arg:"" help:"Table name or ID."`
	RecordIds []string `arg:"" name:"record-ids" help:"Record ID(s) to delete."`
}

// Run deletes record(s).
func (c *RecordsDeleteCmd) Run(globals *Globals) error {
	out, cl, baseID := mustBase(globals, "")
	if cl == nil {
		return nil
	}

	if len(c.RecordIds) == 0 {
		out.Error("VALIDATION_ERROR", "at least one record ID is required")
		return nil
	}

	if len(c.RecordIds) == 1 {
		if err := cl.DeleteRecord(baseID, c.Table, c.RecordIds[0]); err != nil {
			out.Error(errorCode(err), err.Error())
			return nil
		}
		out.Success(map[string]any{"deleted": 1, "ids": c.RecordIds})
	} else {
		if err := cl.DeleteRecords(baseID, c.Table, c.RecordIds); err != nil {
			out.Error(errorCode(err), err.Error())
			return nil
		}
		out.Success(map[string]any{"deleted": len(c.RecordIds), "ids": c.RecordIds})
	}

	return nil
}

// readFieldsInput returns JSON from the flag value or stdin.
func readFieldsInput(flagValue string) (string, error) {
	if flagValue != "" {
		return flagValue, nil
	}

	stat, err := os.Stdin.Stat()
	if err != nil {
		return "", fmt.Errorf("no fields provided; use --fields or pipe JSON via stdin")
	}
	isPipe := (stat.Mode() & os.ModeCharDevice) == 0
	if !isPipe {
		return "", fmt.Errorf("no fields provided; use --fields or pipe JSON via stdin")
	}

	data, err := io.ReadAll(os.Stdin)
	if err != nil {
		return "", fmt.Errorf("read stdin: %w", err)
	}

	trimmed := strings.TrimSpace(string(data))
	if trimmed == "" {
		return "", fmt.Errorf("empty stdin input")
	}

	return trimmed, nil
}
